go/pkg/cas/upload_test.go

```diff
@@ -453,6 +470,185 @@ func TestStreaming(t *testing.T) {
 	}
 }
 
+func TestMissingDirs(t *testing.T) {
+	mustDigest := func(m proto.Message) *repb.Digest {
+		d, err := digest.NewFromMessage(m)
+		if err != nil {
+			t.Fatal(err)
+		}
+		return d.ToProto()
+	}
+
+	barDigest := digest.NewFromBlob([]byte("bar")).ToProto()
+	bazDigest := mustDigest(&repb.Directory{})
+
+	foo := &repb.Directory{
+		Files: []*repb.FileNode{{
+			Name:   "bar",
+			Digest: barDigest,
+		}},
+		Directories: []*repb.DirectoryNode{{
+			Name:   "baz",
+			Digest: bazDigest,
+		}},
+	}
+
+	root := &repb.Directory{
+		Directories: []*repb.DirectoryNode{{
+			Name:   "foo",
+			Digest: mustDigest(foo),
+		}},
+	}
+
+	testCases := []struct {
+		desc     string
+		tree     map[string]*digested
+		wantDirs []missingDir
+	}{
+		{
+			desc: "works",
+			tree: map[string]*digested{
+				"foo/bar": {
```

Comment:
I think it's worth it to have a slightly more complex scenario here too.

Reply:
added a couple more